File: ios/HybridWebmPlayerView.swift
import AVFoundation
import Foundation
import NitroModules
import UIKit

/// Backing view. `layerClass` makes the display layer *be* the view's layer, so
/// there is no second layer to keep in sync with bounds — resizing is handled by
/// UIKit rather than by mirroring frames in `layoutSubviews`.
final class WebmPlayerLayerView: UIView {
  override class var layerClass: AnyClass {
    AVSampleBufferDisplayLayer.self
  }

  var displayLayer: AVSampleBufferDisplayLayer {
    // Safe by construction: layerClass above guarantees the type.
    layer as! AVSampleBufferDisplayLayer
  }
}

/// Presents a player's decoded video.
///
/// The engine is reached by downcasting the `player` prop to the concrete
/// implementation. The prop carries the spec protocol, which deliberately cannot
/// express an `AVSampleBufferDisplayLayer` — that type has no meaning on Android.
final class HybridWebmPlayerView: HybridWebmPlayerViewSpec {
  private let container = WebmPlayerLayerView()

  var view: UIView { container }

  var player: (any HybridWebmPlayerSpec)? {
    didSet {
      // Detach from whatever held the layer before, so a reassigned prop cannot
      // leave two engines enqueuing onto the same surface.
      (oldValue as? HybridWebmPlayer)?.detachDisplayLayer(container.displayLayer)
      (player as? HybridWebmPlayer)?.attachDisplayLayer(container.displayLayer)
    }
  }

  var scaleMode: WebmScaleMode = .contain {
    didSet { applyScaleMode() }
  }

  override init() {
    super.init()
    applyScaleMode()
  }

  deinit {
    (player as? HybridWebmPlayer)?.detachDisplayLayer(container.displayLayer)
  }

  private func applyScaleMode() {
    switch scaleMode {
    case .contain: container.displayLayer.videoGravity = .resizeAspect
    case .cover: container.displayLayer.videoGravity = .resizeAspectFill
    case .stretch: container.displayLayer.videoGravity = .resize
    }
  }
}

File: android/src/main/java/com/margelo/nitro/webmplayer/HybridWebmPlayerView.kt
package com.margelo.nitro.webmplayer

import android.content.Context
import android.view.View
import androidx.media3.common.util.UnstableApi
import androidx.media3.ui.AspectRatioFrameLayout
import androidx.media3.ui.PlayerView

/**
 * Presents a player's decoded video.
 *
 * The ExoPlayer instance is reached by downcasting the `player` prop to the
 * concrete implementation. The prop carries the spec type, which deliberately
 * cannot express a `PlayerView` — that has no meaning on iOS.
 *
 * The predecessor kept a global map of React tags to views and required JS to
 * call `attachToView(tag)`. Here the association is the prop itself, so it
 * cannot outlive the view or bind to the wrong player.
 */
@UnstableApi
class HybridWebmPlayerView(context: Context) : HybridWebmPlayerViewSpec() {

  private val playerView: PlayerView =
    PlayerView(context).apply {
      useController = false
      setShowBuffering(PlayerView.SHOW_BUFFERING_ALWAYS)
      // Not setUseArtwork(false): media3 1.8.0 inverts it, mapping false to
      // ARTWORK_DISPLAY_MODE_FIT, so the deprecated call switched artwork on.
      setArtworkDisplayMode(PlayerView.ARTWORK_DISPLAY_MODE_OFF)
      setShutterBackgroundColor(SHUTTER_COLOR)
    }

  override val view: View
    get() = playerView

  override var player: HybridWebmPlayerSpec? = null
    set(value) {
      // Detach the previous player first, so a reassigned prop cannot leave two
      // players bound to the same surface.
      (field as? HybridWebmPlayer)?.detachSurface(playerView)
      field = value
      (value as? HybridWebmPlayer)?.attachSurface(playerView)
    }

  override var scaleMode: WebmScaleMode = WebmScaleMode.CONTAIN
    set(value) {
      field = value
      playerView.resizeMode = when (value) {
        WebmScaleMode.CONTAIN -> AspectRatioFrameLayout.RESIZE_MODE_FIT
        WebmScaleMode.COVER -> AspectRatioFrameLayout.RESIZE_MODE_ZOOM
        WebmScaleMode.STRETCH -> AspectRatioFrameLayout.RESIZE_MODE_FILL
      }
    }

  private companion object {
    const val SHUTTER_COLOR = 0xFF000000.toInt()
  }
}
